faucet/views.py

```diff
@@ -16,31 +20,45 @@
 class FaucetView(GenericViewSet):
 
     def create(self, request, **kwargs):
-        pub_key = kwargs.get('key')
-        amount = kwargs.get('amount', 100)
+        pub_key = request.data.get('key')
+        amount = request.data.get('amount', 100)
 
-        with redis.lock(f'get_faucet_{pub_key}', expire=120):
+        with redis.lock(f'get_faucet_{pub_key}', timeout=5):
 
             free_tokens = FaucetTransaction.receivable_tokens(pub_key)
             actual_tokens = min(amount, free_tokens)
 
-            config = Config(
-                external_host='epoch:3013',
-                internal_host='epoch:3113',
-                websocket_host='epoch:3114'
-            )
-            client = EpochClient(configs=config)  # connect with the Epoch node
-            try:
-                # check balance
-                balance = client.get_balance()
-                if balance < actual_tokens:
-                    raise ParseError('Faucet is out of cash')
-
-                client.spend(pub_key, actual_tokens)
-            except AException:
-                raise ParseError('Faucet has no account')
-
-            FaucetTransaction.objects.create(
-                public_key=pub_key,
-                amount=actual_tokens
-            )
+            if actual_tokens > 0:
+                epoch_host = settings.EPOCH_HOST
+                config = Config(
+                    external_host=f'{epoch_host}:3013',
+                    internal_host=f'{epoch_host}:3113',
+                    websocket_host=f'{epoch_host}:3114'
+                )
+                client = EpochClient(configs=config)  # connect with the Epoch node
+
+                key_pair = KeyPair.read_from_dir(settings.EPOCH_KEYS, settings.EPOCH_PASSWORD)
+                try:
+                    # check balance
+                    balance = client.get_balance()
+                    if balance < actual_tokens:
+                        raise ParseError('Faucet is out of cash')
+                except AException:
+                    raise ParseError('Faucet has no account')
+
+                try:
+
+                    tx = client.create_transaction(pub_key, int(actual_tokens))
+                    signed_tx = key_pair.sign_transaction(tx)
+                    client.send_signed_transaction(signed_tx)
+                except AException:
+                    raise ParseError(f'Spend TX failed Amount {actual_tokens}')
+
+                FaucetTransaction.objects.create(
+                    public_key=pub_key,
+                    amount=actual_tokens
```

Comment:
we should not mix up tokens and coins. tokens can be created by anyone in a smart contract or similar, while coins have to be mined